File: apps/backend/app/schemas/grading.py
"""Pydantic schemas for avaluacions + qualificacions."""
from __future__ import annotations

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.grading import EstatAvaluacio


# --- Avaluació --------------------------------------------------------------

class AvaluacioBase(BaseModel):
    curs_acad_id: int
    nom: str = Field(min_length=1, max_length=100)
    ordre: int = Field(ge=1, le=20)
    data_inici: date | None = None


class AvaluacioCreate(AvaluacioBase):
    pass


class AvaluacioResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, use_enum_values=True)
    id: int
    curs_acad_id: int
    nom: str
    ordre: int
    estat: EstatAvaluacio
    data_inici: date | None
    data_tancament: date | None


class AvaluacioTransitionRequest(BaseModel):
    target: EstatAvaluacio


# --- Grade matrix (read-side for the spreadsheet) --------------------------

class GradeMatrixAlumne(BaseModel):
    matricula_id: int
    alumne_id: int
    nom: str
    cognoms: str
    dni: str | None
    ralc: str


class GradeMatrixRa(BaseModel):
    id: int
    ordre: int
    codi: str
    descripcio: str
    pes: Decimal


class GradeMatrixCell(BaseModel):
    matricula_id: int
    ra_id: int
    nota: float | None
    comentari: str | None


class GradeMatrixModulCell(BaseModel):
    """Manual mòdul-level override (the final note that bypasses RA mean)."""
    matricula_id: int
    nota: float | None
    comentari: str | None


class GradeMatrixResponse(BaseModel):
    grup_id: int
    modul_id: int
    avaluacio_id: int
    avaluacio_estat: EstatAvaluacio
    can_edit: bool
    alumnes: list[GradeMatrixAlumne]
    ras: list[GradeMatrixRa]
    cells: list[GradeMatrixCell]
    # Manual modul-level overrides for this (modul, avaluacio): empty if none.
    modul_cells: list[GradeMatrixModulCell] = []


# --- Bulk PATCH -------------------------------------------------------------

class QualifRaPatch(BaseModel):
    matricula_id: int
    ra_id: int
    nota: Decimal | None = Field(default=None, ge=0, le=10)
    comentari: str | None = Field(default=None, max_length=2000)


class BulkQualifRaPatch(BaseModel):
    avaluacio_id: int
    patches: list[QualifRaPatch] = Field(min_length=1, max_length=2000)


class QualifRaPatchResult(BaseModel):
    matricula_id: int
    ra_id: int
    ok: bool
    error: str | None = None


class BulkQualifRaPatchResponse(BaseModel):
    results: list[QualifRaPatchResult]
    saved: int
    failed: int


# --- Mòdul-level bulk PATCH (manual final-note overrides) -------------------

class QualifModulPatch(BaseModel):
    matricula_id: int
    nota: Decimal | None = Field(default=None, ge=0, le=10)
    comentari: str | None = Field(default=None, max_length=2000)


class BulkQualifModulPatch(BaseModel):
    avaluacio_id: int
    modul_id: int
    patches: list[QualifModulPatch] = Field(min_length=1, max_length=1000)


class QualifModulPatchResult(BaseModel):
    matricula_id: int
    ok: bool
    error: str | None = None


class BulkQualifModulPatchResponse(BaseModel):
    results: list[QualifModulPatchResult]
    saved: int
    failed: int

File: apps/frontend/src/api/grading.ts
/** Typed client for avaluacions + qualificacions. */
import { api } from "./client";

export type EstatAvaluacio = "oberta" | "docent" | "junta" | "tancada";

export type Avaluacio = {
  id: number;
  curs_acad_id: number;
  nom: string;
  ordre: number;
  estat: EstatAvaluacio;
  data_inici: string | null;
  data_tancament: string | null;
};

export type GradeMatrixAlumne = {
  matricula_id: number;
  alumne_id: number;
  nom: string;
  cognoms: string;
  dni: string | null;
  ralc: string;
};

export type GradeMatrixRa = {
  id: number;
  ordre: number;
  codi: string;
  descripcio: string;
  pes: string;
};

export type GradeMatrixCell = {
  matricula_id: number;
  ra_id: number;
  nota: number | null;
  comentari: string | null;
};

export type GradeMatrixModulCell = {
  matricula_id: number;
  nota: number | null;
  comentari: string | null;
};

export type GradeMatrix = {
  grup_id: number;
  modul_id: number;
  avaluacio_id: number;
  avaluacio_estat: EstatAvaluacio;
  can_edit: boolean;
  alumnes: GradeMatrixAlumne[];
  ras: GradeMatrixRa[];
  cells: GradeMatrixCell[];
  modul_cells: GradeMatrixModulCell[];
};

export type QualifRaPatch = {
  matricula_id: number;
  ra_id: number;
  nota: number | null;
  comentari?: string | null;
};

export type QualifModulPatch = {
  matricula_id: number;
  nota: number | null;
  comentari?: string | null;
};

export type QualifRaPatchResult = {
  matricula_id: number;
  ra_id: number;
  ok: boolean;
  error?: string | null;
};

export type QualifModulPatchResult = {
  matricula_id: number;
  ok: boolean;
  error?: string | null;
};

export type BulkPatchResponse = {
  results: QualifRaPatchResult[];
  saved: number;
  failed: number;
};

export type BulkModulPatchResponse = {
  results: QualifModulPatchResult[];
  saved: number;
  failed: number;
};

export const gradingApi = {
  listAvaluacions: (curs_acad_id?: number) =>
    api<Avaluacio[]>(
      `/avaluacions${curs_acad_id !== undefined ? `?curs_acad_id=${curs_acad_id}` : ""}`,
    ),
  createAvaluacio: (body: {
    curs_acad_id: number;
    nom: string;
    ordre: number;
    data_inici?: string | null;
  }) => api<Avaluacio>("/avaluacions", { method: "POST", body }),
  transition: (id: number, target: EstatAvaluacio) =>
    api<Avaluacio>(`/avaluacions/${id}/transition`, {
      method: "POST",
      body: { target },
    }),

  getGradeMatrix: (params: { grup_id: number; modul_id: number; avaluacio_id: number }) => {
    const sp = new URLSearchParams({
      grup_id: String(params.grup_id),
      modul_id: String(params.modul_id),
      avaluacio_id: String(params.avaluacio_id),
    });
    return api<GradeMatrix>(`/qualificacions/ra?${sp.toString()}`);
  },

  batchPatch: (avaluacio_id: number, patches: QualifRaPatch[]) =>
    api<BulkPatchResponse>("/qualificacions/ra/batch", {
      method: "PATCH",
      body: { avaluacio_id, patches },
    }),

  batchPatchModul: (avaluacio_id: number, modul_id: number, patches: QualifModulPatch[]) =>
    api<BulkModulPatchResponse>("/qualificacions/modul/batch", {
      method: "PATCH",
      body: { avaluacio_id, modul_id, patches },
    }),
};
